draft-vkrasnov-h2-compression-dictionaries.md

```diff
@@ -138,58 +182,87 @@ The SET_COMPRESSION_CONTEXT frame (type=0xTBA).
 ~~~~
 
 The SET_COMPRESSION_CONTEXT frame can be sent by the client on any stream in
-the idle state (prior to the first HEADER frames on the stream).
-It indicates that the stream MAY be compressed by dictionaries with the given
-Context ID.
+the idle state. The frame indicates the compression context ID for the given
+stream. Frames with an assigned context SHALL NOT be compressed using
+dictionaries from a different context. Frames with an assigned context SHALL
+NOT be used as a dictionary for streams with from a different context.
 
 The SET_COMPRESSION_CONTEXT frame contains the following fields:
 
 Context:
-: an 8-bit Context ID that indicates the compression context for the stream.
-A stream can only be compressed by dictionaries with the same Context. If a
-stream SETs a dictionary, it will be assigned the Context indicated by this
-frame. If the frame is ommited, then the context value is assumed to be 0.
-The allowed Context values are 0 through 253.
-A special Context ID of 255 indicates the stream SHALL NOT be compressed at all
-and passed as is.
-A special Context ID of 254 indicates the stream SHALL NOT be compressed with
-any dictionary, and it MUST NOT be used as dictionary for other streams.
+: an 8-bit context ID that indicates the compression context for the stream.
+If the frame is ommited, then the context value is assumed to be 0.
+The allowed context values are 0 through 255.
+A special context ID of 255 indicates the stream SHALL NOT be cross-compressed.
 
 ### The SET_DICTIONARY Frame
 
-The SET_DICTIONARY frame (type=0xTBA).
+The SET_DICTIONARY frame (type=0xTBA) contains one to many Dictionary-Entry.
 
 ~~~~
-+-------------+-------------+
-| Dict ID (8) |   Size (8)  |
-+-------------+-------------+
++---------------+---------------+
+|   Dictionary-Entry (+)    ...
++---------------+---------------+
 ~~~~
 
-The SET_DICTIONARY frame can be sent from the server to the client, on any client
-initiated stream in the open or half-closed (remote) states. The SET_DICTIONARY
-frame MUST precede any DATA frames on that stream. The SET_DICTIONARY frame SHOULD
-be followed by sufficient DATA frames to fill Size octets after decompression,
-even if an RST frame was received for that stream. If not enough DATA was sent,
-the Dictionary for the given ID is considered uninitialized.
+A Dictionary-Entry field is encoded as follows:
+
+~~~~
++-------------------------------+
+|       Dictionary-ID (8)       |
++---+---------------------------+
+| P |        Size (7+)          |
++---+---------------------------+
+| E?|      Truncate? (7+)       |
++---+---------------------------+
+|           Offset? (8+)        |
++-------------------------------+
 
-More than one SET_DICTIONARY frames MAY be sent on a given stream.
+~~~~
 
-The SET_DICTIONARY frame contains the following fields:
+The SET_DICTIONARY frame can be sent from the server to the client, on any
+client initiated stream in the open or half-closed (remote) states. The
```

Comment:
why limit this to "client initiated stream"?

Reply:
right, i actually mentioned it can be used with push, so that is wrong